### ISSUE TITLE
Uint 16 Overflow when fetching very large data from sql server

### ISSUE THREAD

**Issue by ProV1denCEX** [JuliaDatabases/ODBC.jl#226]
The stacktrace shows below: 
```
julia> d_data = ODBC.query("select * from Table_Test")

ERROR: LoadError: BoundsError: attempt to access 1-element Array{UInt16,1} at index [1:868]
Stacktrace:
 [1] throw_boundserror(::Array{UInt16,1}, ::Tuple{UnitRange{Int64}}) at .\abstractarray.jl:484
 [2] checkbounds at .\abstractarray.jl:449 [inlined]
 [3] getindex(::Array{UInt16,1}, ::UnitRange{Int64}) at .\array.jl:737
 [4] cast!(::Type{Union{Missing, String}}, ::ODBC.Query{missing,NamedTuple{(:ts_code, :ann_date, :en
d_date, :type, :p_change_min, :p_change_max, :net_profit_min, :net_profit_max, :last_parent_net, :fi
rst_ann_date, :summary, :change_reason),Tuple{Union{Missing, String},Union{Missing, String},Union{Mi
ssing, String},Union{Missing, String},Union{Missing, Float64},Union{Missing, Float64},Union{Missing,
 Float64},Union{Missing, Float64},Union{Missing, Float64},Union{Missing, String},Union{Missing, Stri
ng},Union{Missing, String}}},Tuple{Array{Union{Missing, String},1},Array{Union{Missing, String},1},A
rray{Union{Missing, String},1},Array{Union{Missing, String},1},Array{Union{Missing, Float64},1},Arra
y{Union{Missing, Float64},1},Array{Union{Missing, Float64},1},Array{Union{Missing, Float64},1},Array
{Union{Missing, Float64},1},Array{Union{Missing, String},1},Array{Union{Missing, String},1},Array{Un
ion{Missing, String},1}}}, ::Int64) at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:212
 [5] (::getfield(ODBC, Symbol("##10#12")))(::Int64) at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src
\Query.jl:104
 [6] foreach(::getfield(ODBC, Symbol("##10#12")), ::UnitRange{Int64}) at .\abstractarray.jl:1835

ERROR: LoadError: BoundsError: attempt to access 1-element Array{UInt16,1} at index [1:868]
Stacktrace:
 [1] throw_boundserror(::Array{UInt16,1}, ::Tuple{UnitRange{Int64}}) at .\abstractarray.jl:484
 [2] checkbounds at .\abstractarray.jl:449 [inlined]
 [3] getindex(::Array{UInt16,1}, ::UnitRange{Int64}) at .\array.jl:737
 [4] cast!(::Type{Union{Missing, String}}, ::ODBC.Query{missing,NamedTuple{(:ts_code, :ann_date, :end_date, :type, :p_change_min, :p_change_max, :net_profit_min, :net_profit_max, :last_parent_net, :first_ann_date, :summary, :change_reason),Tuple{Union{Missing, String},Union{Missing, String},Union{Missing, String},Union{Missing, String},Union{Missing, Float64},Union{Missing, Float64},Union{Missing, Float64},Union{Missing, Float64},Union{Missing, Float64},Union{Missing, String},Union{Missing, String},Union{Missing, String}}},Tuple{Array{Union{Missing, String},1},Array{Union{Missing, String},1},Array{Union{Missing, String},1},Array{Union{Missing, String},1},Array{Union{Missing,
Float64},1},Array{Union{Missing, Float64},1},Array{Union{Missing, Float64},1},Array{Union{Missing, Float64},1},Array{Union{Missing, Float64},1},Array{Union{Missing, String},1},Array{Union{Missing, String},1},Array{Union{Missing, String},1}}}, ::Int64) at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:212
 [5] (::getfield(ODBC, Symbol("##10#12")))(::Int64) at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:104
 [6] foreach(::getfield(ODBC, Symbol("##10#12")), ::UnitRange{Int64}) at .\abstractarray.jl:1835
 [7] ODBC.Query(::ODBC.DSN, ::String) at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:101
 [8] #query#14(::Bool, ::Bool, ::Dict{Int64,Function}, ::Function, ::ODBC.DSN, ::String, ::Type{DataFrames.DataFrame}) at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:278
 [9] query at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:273 [inlined]
 [10] query at F:\Julia\pkgs-1.0.1.1\packages\ODBC\KFiOK\src\Query.jl:264 [inlined]


julia> versioninfo()
Julia Version 1.0.1
Commit 0d713926f8 (2018-09-29 19:05 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, haswell)
Environment:
  JULIA_EDITOR = "F:\Julia\app-1.29.0\atom.exe" -a
  JULIA_NUM_THREADS = 4
  JULIA_PKG_SERVER = https://pkg.juliacomputing.com/
```

The data I tried to fetch has a very large nvarchar key and the size of the data seems to exceed the length of UINT16 (I tried to miss some fields to reduce the query size and ODBC works well).

I guess ODBC.jl may need some solutions for extremely large data? (change UINT16 to 32? 64?)

**Comment by ProV1denCEX** [comment]
ODBC.jl version is 0.8.1

**Comment by openclosure** [comment]
Indeed, I am having an identical issue, only occurring when I try to pull XML fields. Everything else works fine, but pulling that field, whether as a native SQL Server 2016 XML datatype or by casting to `nvarchar(max)`, causes the issue.

**Comment by openclosure** [comment]
Actually, now that I have started over with a fresh repo I am getting a similar error to above but about UInt8. Not sure if I should start a new issue or post it in the other closed issue, but it all feels somewhat related, it seems like data is not getting placed where it should leading to a bounds error when it attempts to cast to julia types. 

I put some print statements into [`Query()` ](https://github.com/JuliaDatabases/ODBC.jl/blob/a0cc3e55b7181a123659428bad4199ceb678100a/src/Query.jl#L101-L104)
```
    jltypes = Type[longtexts[x] ? API.Long{T} : T for (x, T) in enumerate(juliatypes)]
    # Some print statements
    q = Query{rows >= 0 ? rows : missing, NT, typeof(columns)}(stmt, Ref(0), columns, rowsfetched, boundcols, indcols, csizes, types, jltypes)
    if rows != 0
        q.status[] = Int(API.SQLFetchScroll(q.stmt, API.SQL_FETCH_NEXT, 0))
        # The same print statements as above
        q.rowsfetched[] > 0 && foreach(i->cast!(jltypes[i], q, i), 1:length(jltypes))
    else
```

It looks like where `boundcols` would usually be populated with `UInt8` to store the characters, it's not happening for varchar(max).

Example: Trying something like `select 'test'` shows 
`boundcols: Any[UInt8[0x00, 0x00, 0x00, 0x00, 0x00]]` before `SQLFetchScroll` is reached, and 
`boundcols: Any[UInt8[0x74, 0x65, 0x73, 0x74, 0x00]]` after. 

If you try `select cast('test' as varchar(max)`, boundcols shows `Any[UInt8[0x00]]` before and after, a single element, which leads to the bounds error.

**Comment by ProV1denCEX** [comment]
@openclosure Great thanks! But actually, Unicode does matter in my work... Maybe I should try to modify ODBC.jl or try JDBC.jl instead...

**Comment by ProV1denCEX** [comment]
It seems that someone has done this job (to fix this issue) in PR. Maybe we can have a try. @openclosure

**Comment by openclosure** [comment]
I actually tried implementing the fixes made in the PR but I could not get it to work (same error). His fix is for unicode specifically, so I'm not sure it will help my issue of using max, though if I'm understanding correctly it would fix your problem. The PR is only a few lines of edit to Query.jl, you can swap in the function and test it fairly easily. 

Also, I'm not sure what DB you're using but I did test out various windows drivers and had some very limited success with `SQL Server` instead of the generic ODBC driver. 

* `ODBC Driver 13 for SQL Server` - Error
* `SQL Server` - Works but returns garbled data. 
* `SQL Server Native Client 11.0` - Error

**Comment by ProV1denCEX** [comment]
yeah...For me SQL Server Native Client 11.0 and ODBC Driver 13 for SQL Server both work normally with SQL server 2017. Maybe the version of SQL server matters in your issue.

**Comment by openclosure** [comment]
Sorry to hear that. Have you tried the changes made in the PR?

Wondering now if this is related: [Using Large Value Types](https://docs.microsoft.com/en-us/sql/relational-databases/native-client/features/using-large-value-types?view=sql-server-2017#sql-server-native-client-odbc-driver).

It looks like if a varchar or nvarchar over 8000 bytes is returned then `SQL_SS_LENGTH_UNLIMITED` is set to the `ColumnSizePtr` in SQLDescribeCol. Checking the contents of the `csize` variable directly after that SQLDescribeCol API call is made, it is indeed zero, whereas normally it should be the length-in-characters of the data returned. Then later when `internal_allocate` is called, it is called with size zero. Feels like I'm getting closer to the root of the issue.

**Comment by ProV1denCEX** [comment]
Yeah, no more error for me now(temporarily? )

I think you are on the right path 👍 . I gonna have a try on this later

**Comment by openclosure** [comment]
I believe I've identified the problem. A really simple hack along with the inclusion of the commit that fixed #225, fixes things for me. `SQL2Julia` sets the `longtext` flag as true for `SQL_LONGVARCHAR`, but in my situation, using `ODBC Driver 13 for SQL Server` and MS SQL Server 2016, this never returns, as the driver reports long text values by reporting it as a SQL_CHAR and size = 0. 

So for example, this fixes my issue:
```
        alloctypes[x], juliatypes[x], longtexts[x] = API.SQL2Julia[t]
        if (csizes[x] == 0) & (juliatypes[x] == Union{Missing, String})
            longtexts[x] = true
        end
        longtext |= longtexts[x]
```
I'm aware this is probably not a "good" fix. Unfortunately I'm not too sure what needs to be done to make it more general. I will try to work on it and put a PR in though.

**Comment by MegaByte** [comment]
I am having a similar problem loading large varchar fields using the Amazon Athena connector.

```
BoundsError: attempt to access 256-element Array{UInt8,1} at index [1:2384]

Stacktrace:
 [1] throw_boundserror(::Array{UInt8,1}, ::Tuple{UnitRange{Int64}}) at ./abstractarray.jl:484
 [2] checkbounds at ./abstractarray.jl:449 [inlined]
 [3] getindex(::Array{UInt8,1}, ::UnitRange{Int64}) at ./array.jl:735
 [4] cast!(::Type{Union{Missing, String}}, ::ODBC.Query{missing,NamedTuple{(:metadata,),Tuple{Union{Missing, String}}},Tuple{Array{Union{Missing, String},1}}}, ::Int64) at /Users/aaron/.julia/packages/ODBC/KFiOK/src/Query.jl:212
 [5] (::getfield(ODBC, Symbol("##10#12")))(::Int64) at /Users/aaron/.julia/packages/ODBC/KFiOK/src/Query.jl:104
 [6] foreach(::getfield(ODBC, Symbol("##10#12")), ::UnitRange{Int64}) at ./abstractarray.jl:1866
 [7] ODBC.Query(::ODBC.DSN, ::String) at /Users/aaron/.julia/packages/ODBC/KFiOK/src/Query.jl:104
 [8] #query#14(::Bool, ::Bool, ::Dict{Int64,Function}, ::Function, ::ODBC.DSN, ::String, ::Type{DataFrame}) at /Users/aaron/.julia/packages/ODBC/KFiOK/src/Query.jl:278
 [9] query at /Users/aaron/.julia/packages/ODBC/KFiOK/src/Query.jl:273 [inlined]
 [10] query(::ODBC.DSN, ::String) at /Users/aaron/.julia/packages/ODBC/KFiOK/src/Query.jl:264
 [11] top-level scope at In[3]:1
```

**Comment by jekelsc** [comment]
Found the root cause.

I query several SQL Server databases for work with very lazy schema - every string field is nvarchar(max).

ODBC.jl v0.8.2 works fine for these when using Progress Data Direct drivers. For Microsoft ODBC drivers, the BoundsError is usually thrown on the nvarchar(max) fields.

The if statement addition in the March 6 comment by openclosure gets the Microsoft ODBC driver partially working (for large vchar but probably not large varbinary) since MSFT does return size 0 for large fields.

With the if statement in place, any select query where the large varchar fields are placed in the select list after other types work fine. Placing a large varchar field before a field of a "normal" type will cause failure.

Checking the SQLGetInfo return of SQL_GETDATA_EXTENSIONS for the MSFT driver reveals no support for SQL_GD_ANY_COLUMN. This means: "SQLGetData is required to return data only from unbound columns that occur after the last bound column, are called in order of increasing column number, and are not in a row in a block of rows."

The ODBC.Query function calls SQLBindCol on all columns that are not longtexts, regardless of the order they appear, so the logic only works with drivers that support SQL_GD_ANY_COLUMN or if one puts all large fields at the end of the field list for drivers that don't.

Looking at the pyodbc code, it avoids this issue completely by using SQLGetData for every field type instead of binding some. I found no calls to SQLBindCol.

This description should help unblock some folks as there may be some workarounds if you can change your query (putting large fields at the end of the list or casting them to a specific varchar size).

I have no Pull Request ready as this may involve a bit of a rewrite and heavy testing of the main query path. 

Whether or not the data binding logic should vary depending on the driver capabilities returned from SQLGetInfo for SQL_GETDATA_EXTENSIONS is an open question. 

If you are curious about the details of tracking this down, after the first failing SQLGetData in cast! for API.Long types, adding some diagnostic code to call SQLGetDiagRec reported error code 07009, "Invalid descriptor index," which lead to several helpful hints about previously bound columns.

**Comment by quinnj** [comment]
Thank you @jekelsc for the detailed investigation and report! I think it would make sense to change the "longtext" [case](https://github.com/JuliaDatabases/ODBC.jl/blob/a249dbfd22563137400ebc4d6f7e409114aacf50/src/Query.jl#L75) to just call SQLGetData for every column. I can work on this in a couple of days, but if anyone else wants to take a crack at it, feel free! I'd be more than happy to review a PR and help answer any questions if needed.

**Comment by jekelsc** [comment]
@quinnj Just letting you know I am working on this. I have working code with minimal changes that I hope to package up in a PR for you tomorrow morning.
